File: packages/impress-mcp/src/raster.ts
/**
 * PDF page rasterisation, with no npm dependency.
 *
 * MCP image content is raster only, so a PDF can never be returned directly.
 * The target workflow — the user on a phone, the agent on the Mac, the
 * conversation as the only display surface — needs the page as a PNG or the
 * result is unviewable.
 *
 * Rather than pull in a PDF renderer (pdfjs + canvas is tens of megabytes and
 * a native build), this shells out to `osascript -l JavaScript`, which is part
 * of macOS and has a bridge to PDFKit/AppKit. The whole suite is macOS-native,
 * and the MCP server always runs on the user's Mac even when imbib itself is
 * remote (the phone), so the bytes are always rasterisable locally.
 *
 * Every failure path returns `{ ok: false }` — never throws — so callers can
 * degrade to text plus an on-disk path instead of losing the tool call.
 */

import { execFile } from "child_process";
import { promisify } from "util";
import { mkdir, writeFile, readFile, unlink } from "fs/promises";
import { tmpdir } from "os";
import { join } from "path";

const execFileAsync = promisify(execFile);

/** Longest edge, in pixels, of a rendered page. ~1100px stays readable for
 * body text while keeping a typical page well under MAX_INLINE_IMAGE_BYTES. */
export const DEFAULT_PAGE_MAX_DIM = 1100;

/** Where rasterised inputs are parked so captions can quote a real path. */
export function rasterScratchDir(): string {
  return join(tmpdir(), "impress-mcp");
}

export type RasterSuccess = {
  ok: true;
  /** Raw base64 PNG, no data: URI prefix. */
  base64: string;
  /** Decoded PNG size in bytes. */
  byteLength: number;
  pageCount: number;
  /** 1-based page actually rendered (clamped into range). */
  page: number;
  width: number;
  height: number;
  /** The source PDF on disk — useful to an agent that can open files. */
  pdfPath: string;
};

export type RasterFailure = {
  ok: false;
  error: string;
  pdfPath?: string;
};

export type RasterResult = RasterSuccess | RasterFailure;

/**
 * JXA: open the PDF with PDFKit, render one page to PNG, report the page
 * count. `boundsForBox(0)` is kPDFDisplayBoxMediaBox; `4` is
 * NSBitmapImageFileTypePNG — the named constants are not on the JXA bridge.
 */
const RASTER_JXA = `
function run(argv) {
  ObjC.import('Foundation');
  ObjC.import('AppKit');
  ObjC.import('Quartz');
  var inPath = argv[0], outPath = argv[1];
  var pageArg = parseInt(argv[2], 10) || 1;
  var maxDim = parseInt(argv[3], 10) || 1100;

  var doc = $.PDFDocument.alloc.initWithURL($.NSURL.fileURLWithPath($(inPath)));
  if (!doc || doc.isNil()) return JSON.stringify({ error: 'not a readable PDF' });
  var count = parseInt(doc.pageCount, 10);
  if (!count) return JSON.stringify({ error: 'PDF has no pages' });
  var idx = pageArg < 1 ? 1 : (pageArg > count ? count : pageArg);
  var page = doc.pageAtIndex(idx - 1);
  if (!page || page.isNil()) return JSON.stringify({ error: 'page ' + idx + ' not readable' });

  var bounds = page.boundsForBox(0);
  var w = bounds.size.width, h = bounds.size.height;
  if (!w || !h) return JSON.stringify({ error: 'page has zero bounds' });
  var scale = Math.min(maxDim / w, maxDim / h, 4);
  if (!isFinite(scale) || scale <= 0) scale = 1;
  var tw = Math.max(1, Math.round(w * scale)), th = Math.max(1, Math.round(h * scale));

  var img = page.thumbnailOfSizeForBox({ width: tw, height: th }, 0);
  var rep = $.NSBitmapImageRep.imageRepWithData(img.TIFFRepresentation);
  var png = rep.representationUsingTypeProperties(4, $());
  if (!png || png.isNil()) return JSON.stringify({ error: 'PNG encoding failed' });
  png.writeToFileAtomically($(outPath), true);
  return JSON.stringify({ pageCount: count, page: idx, width: tw, height: th });
}
`;

/**
 * Render one page of a PDF to PNG.
 *
 * @param pdf   the PDF bytes (as returned by `response.arrayBuffer()`)
 * @param opts.page     1-based page number, clamped to the document (default 1)
 * @param opts.maxDim   longest edge in pixels (default DEFAULT_PAGE_MAX_DIM)
 * @param opts.basename slug used for the scratch filenames
 */
export async function rasterizePDFPage(
  pdf: ArrayBuffer | Uint8Array,
  opts: { page?: number; maxDim?: number; basename?: string } = {}
): Promise<RasterResult> {
  const page = Math.max(1, Math.floor(opts.page ?? 1));
  const maxDim = opts.maxDim ?? DEFAULT_PAGE_MAX_DIM;
  const slug = (opts.basename ?? "document").replace(/[^A-Za-z0-9._-]/g, "_").slice(0, 64);

  const dir = rasterScratchDir();
  const pdfPath = join(dir, `${slug}.pdf`);
  const pngPath = join(dir, `${slug}-p${page}.png`);

  try {
    await mkdir(dir, { recursive: true });
    const bytes = pdf instanceof Uint8Array ? pdf : new Uint8Array(pdf);
    await writeFile(pdfPath, bytes);
  } catch (error) {
    return { ok: false, error: `could not stage the PDF: ${messageOf(error)}` };
  }

  let meta: { pageCount?: number; page?: number; width?: number; height?: number; error?: string };
  try {
    const { stdout } = await execFileAsync(
      "osascript",
      ["-l", "JavaScript", "-e", RASTER_JXA, pdfPath, pngPath, String(page), String(maxDim)],
      { timeout: 30_000 }
    );
    meta = JSON.parse(stdout.trim() || "{}");
  } catch (error) {
    // Missing osascript (non-macOS host) or a JXA error both land here.
    return { ok: false, error: `PDF rasterisation failed: ${messageOf(error)}`, pdfPath };
  }

  if (meta.error) return { ok: false, error: meta.error, pdfPath };

  try {
    const png = await readFile(pngPath);
    await unlink(pngPath).catch(() => {});
    return {
      ok: true,
      base64: png.toString("base64"),
      byteLength: png.byteLength,
      pageCount: meta.pageCount ?? 1,
      page: meta.page ?? page,
      width: meta.width ?? 0,
      height: meta.height ?? 0,
      pdfPath,
    };
  } catch (error) {
    return { ok: false, error: `rendered page could not be read back: ${messageOf(error)}`, pdfPath };
  }
}

function messageOf(error: unknown): string {
  if (error instanceof Error) return error.message.trim();
  return String(error);
}

File: crates/impress-mcp/src/raster.rs
//! PDF page rasterisation, with no new dependency.
//!
//! MCP image content is raster only, so a PDF can never be returned directly.
//! The workflow that needs this — the user on a phone, the agent on the Mac,
//! the conversation as the only display surface — needs the page as a PNG or
//! the result is unviewable. Before this, `get_pdf` could only report a path
//! and a byte count, which is useless to anyone not sitting at that Mac.
//!
//! Rather than pull in a PDF renderer (pdfium plus a native build, tens of
//! megabytes), this shells out to `osascript -l JavaScript`, which ships with
//! macOS and bridges to PDFKit and AppKit. The whole suite is macOS-native and
//! this binary always runs on the user's Mac, so the trade is a good one.
//!
//! Ported from `packages/impress-mcp/src/raster.ts` during the TypeScript
//! retirement; the JXA below is that file's, unchanged, because it was already
//! doing the right thing.
//!
//! Spawning a subprocess is fine *here*: this is a standalone CLI. The
//! no-`Process()` invariant that binds impel applies to the sandboxed app, not
//! to this binary.

use std::path::{Path, PathBuf};
use std::process::Command;

/// Longest edge, in pixels, of a rendered page. ~1100px keeps body text legible
/// while keeping a typical page small enough to inline.
pub const DEFAULT_PAGE_MAX_DIM: u32 = 1100;

/// JXA: open the PDF with PDFKit, render one page to PNG, report the page
/// count. `boundsForBox(0)` is kPDFDisplayBoxMediaBox; `4` is
/// NSBitmapImageFileTypePNG — the named constants are not on the JXA bridge.
const RASTER_JXA: &str = r#"
function run(argv) {
  ObjC.import('Foundation');
  ObjC.import('AppKit');
  ObjC.import('Quartz');
  var inPath = argv[0], outPath = argv[1];
  var pageArg = parseInt(argv[2], 10) || 1;
  var maxDim = parseInt(argv[3], 10) || 1100;

  var doc = $.PDFDocument.alloc.initWithURL($.NSURL.fileURLWithPath($(inPath)));
  if (!doc || doc.isNil()) return JSON.stringify({ error: 'not a readable PDF' });
  var count = parseInt(doc.pageCount, 10);
  if (!count) return JSON.stringify({ error: 'PDF has no pages' });
  var idx = pageArg < 1 ? 1 : (pageArg > count ? count : pageArg);
  var page = doc.pageAtIndex(idx - 1);
  if (!page || page.isNil()) return JSON.stringify({ error: 'page ' + idx + ' not readable' });

  var bounds = page.boundsForBox(0);
  var w = bounds.size.width, h = bounds.size.height;
  if (!w || !h) return JSON.stringify({ error: 'page has zero bounds' });
  var scale = Math.min(maxDim / w, maxDim / h, 4);
  if (!isFinite(scale) || scale <= 0) scale = 1;
  var tw = Math.max(1, Math.round(w * scale)), th = Math.max(1, Math.round(h * scale));

  var img = page.thumbnailOfSizeForBox({ width: tw, height: th }, 0);
  var rep = $.NSBitmapImageRep.imageRepWithData(img.TIFFRepresentation);
  var png = rep.representationUsingTypeProperties(4, $());
  if (!png || png.isNil()) return JSON.stringify({ error: 'PNG encoding failed' });
  png.writeToFileAtomically($(outPath), true);
  return JSON.stringify({ pageCount: count, page: idx, width: tw, height: th });
}
"#;

/// A rendered page.
pub struct Raster {
    /// Raw PNG bytes.
    pub png: Vec<u8>,
    pub page_count: u32,
    /// 1-based page actually rendered, clamped into range.
    pub page: u32,
    pub width: u32,
    pub height: u32,
}

/// Where rasterised output is parked, so a caption can quote a real path.
pub fn scratch_dir() -> PathBuf {
    std::env::temp_dir().join("impress-mcp")
}

/// Render one page of a PDF on disk to PNG.
///
/// Never panics and never returns a partial success: every failure path yields
/// `Err(reason)` so the caller can degrade to text plus the path rather than
/// losing the tool call.
pub fn rasterize_pdf_page(pdf_path: &Path, page: u32, max_dim: u32) -> Result<Raster, String> {
    if !pdf_path.exists() {
        return Err(format!("no PDF at {}", pdf_path.display()));
    }
    let dir = scratch_dir();
    std::fs::create_dir_all(&dir)
        .map_err(|e| format!("could not create {}: {e}", dir.display()))?;

    let slug: String = pdf_path
        .file_stem()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| "document".into())
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '.' || c == '-' {
                c
            } else {
                '_'
            }
        })
        .take(64)
        .collect();
    let png_path = dir.join(format!("{slug}-p{page}.png"));

    let page = page.max(1);
    let max_dim = if max_dim == 0 {
        DEFAULT_PAGE_MAX_DIM
    } else {
        max_dim
    };

    let output = Command::new("osascript")
        .args([
            "-l",
            "JavaScript",
            "-e",
            RASTER_JXA,
            &pdf_path.to_string_lossy(),
            &png_path.to_string_lossy(),
            &page.to_string(),
            &max_dim.to_string(),
        ])
        .output()
        .map_err(|e| format!("could not run osascript: {e}"))?;

    if !output.status.success() {
        return Err(format!(
            "osascript failed: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        ));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let meta: serde_json::Value =
        serde_json::from_str(stdout.trim()).map_err(|e| format!("unreadable JXA output: {e}"))?;
    if let Some(err) = meta.get("error").and_then(|e| e.as_str()) {
        return Err(err.to_string());
    }

    let png = std::fs::read(&png_path)
        .map_err(|e| format!("could not read {}: {e}", png_path.display()))?;
    // Best-effort cleanup: the PNG has been read into memory, and leaving
    // scratch files behind across a long session adds up.
    let _ = std::fs::remove_file(&png_path);

    let u = |k: &str| meta.get(k).and_then(|v| v.as_u64()).unwrap_or(0) as u32;
    Ok(Raster {
        png,
        page_count: u("pageCount"),
        page: u("page"),
        width: u("width"),
        height: u("height"),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_missing_pdf_is_an_error_not_a_panic() {
        let err = rasterize_pdf_page(Path::new("/nonexistent/nope.pdf"), 1, 800).unwrap_err();
        assert!(err.contains("no PDF at"), "got {err}");
    }

    #[test]
    fn a_non_pdf_is_rejected_by_pdfkit() {
        // Only meaningful on macOS, where osascript exists.
        if Command::new("osascript")
            .arg("-e")
            .arg("1")
            .output()
            .is_err()
        {
            return;
        }
        let dir = scratch_dir();
        std::fs::create_dir_all(&dir).unwrap();
        let fake = dir.join("not-a-pdf-test.pdf");
        std::fs::write(&fake, b"this is not a PDF").unwrap();
        let err = rasterize_pdf_page(&fake, 1, 400).unwrap_err();
        let _ = std::fs::remove_file(&fake);
        assert!(!err.is_empty(), "expected a reason");
    }
}
